routes/runtime/github.rb

```diff
@@ -0,0 +1,128 @@
+      # list.
+      scopes = [runner.workflow_job&.dig("head_branch"), repository.default_branch].compact
+      entry = repository.cache_entries_dataset
+        .exclude(committed_at: nil)
+        .where(key: keys, version: version, scope: scopes).all
+        .min_by { keys.index(_1.key) + (scopes.index(_1.scope) * keys.size) }
+
+      fail CloverError.new(204, "NotFound", "No cache entry") if entry.nil?
+
+      entry.update(last_accessed_at: Time.now, last_accessed_by: runner.id)
+      signed_url = repository.url_presigner.presigned_url(:get_object, bucket: repository.bucket_name, key: entry.blob_key, expires_in: 900)
+
+      {
+        scope: entry.scope,
+        cacheKey: entry.key,
+        cacheVersion: entry.version,
+        creationTime: entry.created_at,
+        archiveLocation: signed_url
+      }
+    end
+
+    r.on "caches" do
+      # listCache
+      r.get true do
+        key = r.params["key"]
+        fail CloverError.new(204, "NotFound", "No cache entry") if key.nil?
+
+        scopes = [runner.workflow_job&.dig("head_branch"), repository.default_branch].compact
+        entries = repository.cache_entries_dataset
+          .exclude(committed_at: nil)
+          .where(key: key, scope: scopes).all
+
+        {
+          totalCount: entries.count,
+          artifactCaches: entries.map do
+            {
+              scope: _1.scope,
+              cacheKey: _1.key,
+              cacheVersion: _1.version,
+              creationTime: _1.created_at
+            }
+          end
+        }
+      end
+
+      repository.setup_blob_storage unless repository.access_key
+
+      # reserveCache
+      r.post true do
+        key = r.params["key"]
+        version = r.params["version"]
+        size = r.params["cacheSize"].to_i
+        fail CloverError.new(400, "InvalidRequest", "Wrong parameters") if key.nil? || version.nil? || size == 0
+
+        unless (scope = runner.workflow_job&.dig("head_branch"))
+          # YYYY: If the webhook not delivered yet, we can try to get the branch from the API
+          Clog.emit("The runner does not have a workflow job") { {no_workflow_job: {ubid: runner.ubid, repository_ubid: repository.ubid}} }
+          fail CloverError.new(400, "InvalidRequest", "No workflow job data available")
+        end
+
+        entry = GithubCacheEntry.create_with_id(repository_id: runner.repository.id, key: key, version: version, size: size, scope: scope, created_by: runner.id)
+
+        upload_id = repository.blob_storage_client.create_multipart_upload(bucket: repository.bucket_name, key: entry.blob_key).upload_id
+        entry.update(upload_id: upload_id)
+
+        max_chunk_size = 32 * 1024 * 1024 # 32MB
+        presigned_urls = (1..size.fdiv(max_chunk_size).ceil).map do
+          repository.url_presigner.presigned_url(:upload_part, bucket: repository.bucket_name, key: entry.blob_key, upload_id: upload_id, part_number: _1, expires_in: 900)
```

Comment:
Does this leave 15 minutes to start he upload or to finish the upload?

Reply:
It leaves 15 minutes to start upload. Once the connection is open, the expiration doesn't matter. However, since we upload in chunks, we need to start uploading all chunks before the expiration.